release/designs/artifact-table-generation.md

```diff
@@ -0,0 +1,66 @@
+
+### Tenets
+
+* **Simple:** Adding another component to the bundle should be simple and with minimal changes and little to no code duplication
+* **Declarative:** The list of artifacts for a project should be stored declaratively, for abstraction and easier tracking
+
+## Goals
+
+As an EKS Anywhere developer, I want to
+
+* Make it easy to add another component to the bundle
+* Make the review process for bundle changes and refactors easier
+* Reduce code duplication and techdebt in the release tooling
+
+## Implementation details
+
+Currently, the artifacts declaration for each project is stored in its separate Go file, along with the logic for packaging assets (configuring source/release destinations, versioning, etc), that is mostly common across all these files.
+
+Instead of this, it would be ideal to take the declarative approach and define the artifacts for a project as a set of Go custom structs in a single file. This file would contain the list of artifacts (images, manifests, archives) for each component and some other parameters that are required by the common methods, which will all be extracted into a single file. During the release process, we can then iterate over these structs and feed them into the "oracle" that contains the business logic to get the final artifacts table.
+
+**Current folder structure:**
+
+```
+   release
+      |_____ pkg
+              |_______ assets_capi.go
+              |_______ assets_capa.go
+              |_______ assets_capd.go
+              |_______ assets_capv.go
+              .
+              .
+              .
+```
+
+**Proposed folder structure:**
+
+```
+   release
+      |_____ pkg
+              |_______ assets
+                         |_______ assets.go # contains processing logic for artifacts
+                         |_______ config.go # contains declarative definition for project artifacts
+```
+
+The following schema will be used to define project artifacts as a Go struct:
+
+```Go
+type AssetConfig struct {
+	ProjectName     string               `json:"projectName,omitempty"`
+	ProjectPath     string               `json:"projectPath,omitempty"`
+	Archives        []*Archive           `json:"archives,omitempty"`
+	ArchivePrefix   string               `json:"archivePrefix,omitempty"`
+	Images          []*Image             `json:"images,omitempty"`
+	ImageRepoPrefix string               `json:"imageRepoPrefix,omitempty"`
+	ImageTagOptions []string             `json:"imageTagOptions,omitempty"`
+	Manifests       []*ManifestComponent `json:"manifests,omitempty"`
+	NoGitTag        bool                 `json:"noGitTag,omitempty"`
```

Comment:
Is this also kind of an option?

Reply:
Yes but there is just one project which doesn't have - the bottlerocket bootstrap. We might add more in the future so kept it configurable